### PR TITLE
Optimise IO & add ELK

### PR DESCRIPTION
**Pull request by mdbudd**

### CODE REVIEW

**Review by baldygeezer — CHANGES_REQUESTED** [review]
Looks good. At the moment the composer install step happens in the build script: /bin/dev-build
That will also need changing.

File: Dockerfile [hunk 1]
```diff
@@ -16,6 +16,9 @@ RUN docker-php-ext-install zip
 
 ENV COMPOSER_ALLOW_SUPERUSER=1
 
+COPY ./composer.json ./
+RUN composer install
```
Comment (bartfeenstra) [inline Dockerfile:20]: The application must not be installed into the container image for dev environments. That should (in the near future) only be done for deployments where we do not need to do any development work (e.g. prod)

File: docker-compose.yml [hunk 2]
```diff
@@ -9,6 +9,7 @@ services:
       dockerfile: Dockerfile
     volumes:
       - .:/srv/ioda
+      - /srv/ioda/vendor
```
Comment (bartfeenstra) [inline docker-compose.yml:12]: Is that a correct volume definition? As far as I understand Docker, each volume definition should have at least two 'fields' within the string value, separated by a colon.

**Review by bartfeenstra — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 2 above)
Comment (mdbudd) [reply docker-compose.yml:12]: Yes it is, you don't have to have two fields, you can have a direct bind.

**Review by AndreaLDavies — APPROVED** [review]

File: docker-compose.yml (hunk 2 above)
Comment (bartfeenstra) [reply docker-compose.yml:31]: What does it bind to in such cases? Same path on the host as inside the container?

File: docker-compose.yml [hunk 3]
```diff
@@ -9,7 +28,222 @@ services:
...
+      test:
+        [
+          "CMD-SHELL",
+          "curl -s -I http://localhost:5601 | grep -q 'HTTP/1.1 302 Found'",
+        ]
+      interval: 10s
+      timeout: 10s
+      retries: 120
+
+  metricbeat01:
```
Comment (bartfeenstra) [inline docker-compose.yml:178]: This service should get a more descriptive name that explains what the service's purpose is.

File: docker-compose.yml [hunk 4]
```diff
@@ -9,7 +28,222 @@ services:
...
+      - "/var/lib/docker/containers:/var/lib/docker/containers:ro"
+      - "/var/run/docker.sock:/var/run/docker.sock:ro"
+    environment:
+      - ELASTIC_USER=elastic
+      - ELASTIC_PASSWORD=${ELASTIC_PASSWORD}
+      - ELASTIC_HOSTS=https://es01:9200
+      - KIBANA_HOSTS=http://kibana:5601
+      - LOGSTASH_HOSTS=http://logstash01:9600
+
+  logstash01:
```
Comment (bartfeenstra) [inline docker-compose.yml:223]: Do we expect additional Logstash services?

File: docker-compose.yml [hunk 5]
```diff
@@ -9,7 +28,222 @@ services:
...
+        condition: service_healthy
+    image: docker.elastic.co/logstash/logstash:${STACK_VERSION}
+    labels:
+      co.elastic.logs/module: logstash
+    user: root
+    volumes:
+      - certs:/usr/share/logstash/certs
+      - logstashdata01:/usr/share/logstash/data
+      - "./elk/logstash_ingest_data/:/usr/share/logstash/ingest_data/"
+      # - "./elk/logstash.conf:/usr/share/logstash/pipeline/logstash.conf:ro"
```
Comment (bartfeenstra) [inline docker-compose.yml:237]: No commented-out code.

File: docker-compose.yml [hunk 6]
```diff
@@ -9,7 +28,222 @@ services:
...
+      - "./elk/logstash/drv:/usr/share/logstash/drv/"
+      - "./elk/logstash/pipelines:/usr/share/logstash/bin/conf.d/"
+      # - "./elk/logstash.yml:/usr/share/logstash/config/logstash.yml:ro"
```
Comment (bartfeenstra) [inline docker-compose.yml:240]: No commented-out code.

File: docker-compose.yml [hunk 7]
```diff
@@ -9,7 +28,222 @@ services:
...
+      - "./elk/pipelines.yml:/usr/share/logstash/config/pipelines.yml:ro"
+    environment:
+      - xpack.monitoring.enabled=false
+      - ELASTIC_USER=elastic
+      - ELASTIC_PASSWORD=${ELASTIC_PASSWORD}
+      - SQL_USER=${SQL_USER}
+      - SQL_PASSWORD=${SQL_PASSWORD}
+      - ELASTIC_HOSTS=https://es01:9200
+#    network_mode: host
```
Comment (bartfeenstra) [inline docker-compose.yml:249]: No commented-out code.

File: elk/filebeat_ingest_data/Air_Quality.log
Comment (bartfeenstra) [inline elk/filebeat_ingest_data/Air_Quality.log:1]: Should this file be in our repo?

File: elk/logstash/drv/mssql-jdbc-12.2.0.jre8.jar
Comment (bartfeenstra) [inline elk/logstash/drv/mssql-jdbc-12.2.0.jre8.jar:1]: Let's remove this file from our repo, and download it during the build process instead.

File: elk/pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+#
+#   If using dead_letter_queue.enable: true, the directory path where the data files will be stored.
+#   Default is path.data/dead_letter_queue
+#
+#   path.dead_letter_queue:
+
+- pipeline.id: DEV_ioda_metrics.conf
+  path.config: "/usr/share/logstash/bin/conf.d/DEV_ioda_metrics.conf"
+# - pipeline.id: DEV_ioda_awards.conf
```
Comment (bartfeenstra) [inline elk/pipelines.yml:116]: No unused code.

File: parse.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+var parser = require("liqe");
+let query = process.argv[2].slice(0, -1).replace(/\r?\n|\r/g, "")
+// console.log(parser.parse(query))
```
Comment (bartfeenstra) [inline parse.js:3]: No unused code.

File: parse.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+console.log("out")
+// query = "doe AND 1212 AND !foo"
```
Comment (bartfeenstra) [inline parse.js:5]: No unused code.

File: parse.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+let out = JSON.stringify(parser.parse(query))
+console.log(out)
```
Comment (bartfeenstra) [inline parse.js:7]: No debugging code.

File: parse.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+var parser = require("liqe");
+let query = process.argv[2].slice(0, -1).replace(/\r?\n|\r/g, "")
+// console.log(parser.parse(query))
+console.log("out")
```
Comment (bartfeenstra) [inline parse.js:4]: No debugging code.

File: server.js
Comment (bartfeenstra) [inline server.js:1]: What is this file used for?

File: server.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+app.use(express.static('openapi'));
+
+app.get('/', (req, res) => {
+    res.send('Hello IODA!');
+});
+
+app.get('/parse', (req, res) => {
+    let query = req.query
+    // query = query.q.slice(0, -1).replace(/\r?\n|\r/g, "")
```
Comment (bartfeenstra) [inline server.js:14]: No unused code.

File: src/Controller/LuceneController.php
Comment (bartfeenstra) [inline src/Controller/LuceneController.php:1]: Is this controller a test fixture?

File: src/Controller/LuceneController.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+     * @throws TransportExceptionInterface
+     */
+    #[Route('/parse/{query}', name: 'luce_index', defaults: ['query' => 'name:foo AND NOT (bio:bar OR bio:baz) AND height:[100 TO 200]'], methods: ['GET'])]
+    public function luce(
+        HttpClientInterface $httpClient,
+        Request $request,
+        LoggerInterface $logger,
+        string $query
+    ): Response {
+        $response = $httpClient->request('GET', 'http://node:8081/parse?q=' . $query);
```
Comment (bartfeenstra) [inline src/Controller/LuceneController.php:31]: Let's put the bit where we have a Lucene query string parsed in an API function, and add an integration test for it.

File: docker-compose.yml (hunk 2 above)
Comment (bartfeenstra) [reply docker-compose.yml:33]: Can you link the specific docs for this syntax? Can't seem to find it myself.

**Comment by bartfeenstra** [comment]
I added https://github.com/nihruk/console/issues/80 as a follow-up.

**Review by bartfeenstra — APPROVED** [review]

**Pull request merged**